core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java

```diff
@@ -0,0 +1,405 @@
+  /** The worker configuration checker. */
+  private final ServerConfigurationChecker mWorkerConfigChecker = new ServerConfigurationChecker();
+
+  /** The hostname of this master. */
+  private String mMasterHostname = Configuration.get(PropertyKey.MASTER_HOSTNAME);
+
+  /** The master ID for this master. */
+  private AtomicReference<Long> mMasterId = new AtomicReference<>(-1L);
+
+  /** Client for all meta master communication. */
+  private final MetaMasterMasterClient mMetaMasterClient
+      = new MetaMasterMasterClient(MasterClientConfig.defaults());
+
+  /** The start time for when the master started serving the RPC server. */
+  private long mStartTimeMs = -1;
+
+  /**
+   * Creates a new instance of {@link DefaultMetaMaster}.
+   *
+   * @param blockMaster a block master handle
+   * @param masterContext the context for Alluxio master
+   */
+  DefaultMetaMaster(BlockMaster blockMaster, MasterContext masterContext) {
+    this(blockMaster, masterContext, ExecutorServiceFactories
+        .fixedThreadPoolExecutorServiceFactory(Constants.META_MASTER_NAME, 2));
+  }
+
+  /**
+   * Creates a new instance of {@link DefaultMetaMaster}.
+   *
+   * @param blockMaster a block master handle
+   * @param masterContext the context for Alluxio master
+   * @param executorServiceFactory a factory for creating the executor service to use for running
+   *        maintenance threads
+   */
+  DefaultMetaMaster(BlockMaster blockMaster, MasterContext masterContext,
+      ExecutorServiceFactory executorServiceFactory) {
+    super(masterContext, new SystemClock(), executorServiceFactory);
+    mIsInSafeMode = masterContext.getmSafeModeManager().isInSafeMode();
+    mBlockMaster = blockMaster;
+    mBlockMaster.registerLostWorkerFoundListener(this::lostWorkerFoundHandler);
+    mBlockMaster.registerWorkerLostListener(this::workerLostHandler);
+    mBlockMaster.registerNewWorkerConfListener(this::registerNewWorkerConfHandler);
+  }
+
+  @Override
+  public Map<String, TProcessor> getServices() {
+    Map<String, TProcessor> services = new HashMap<>();
+    services.put(Constants.META_MASTER_CLIENT_SERVICE_NAME,
+        new MetaMasterClientService.Processor<>(new MetaMasterClientServiceHandler(this)));
+    services.put(Constants.META_MASTER_MASTER_SERVICE_NAME,
+        new MetaMasterMasterService.Processor<>(new MetaMasterMasterServiceHandler(this)));
+    return services;
+  }
+
+  @Override
+  public String getName() {
+    return Constants.META_MASTER_NAME;
+  }
+
+  @Override
+  public Set<Class<? extends Server>> getDependencies() {
+    return DEPS;
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    throw new IOException(ExceptionMessage.UNEXPECTED_JOURNAL_ENTRY.getMessage(entry));
+  }
+
+  @Override
+  public void resetState() {}
+
+  @Override
+  public Iterator<JournalEntry> getJournalEntryIterator() {
+    return null;
```

Comment:
this might cause an NPE. Better to return an empty iterator

Reply:
Changed to ```return Iterators.emptyIterator();``` to avoid NPE, thanks!